### PR TITLE
Add User Token Grant flow

### PR DESCRIPTION
**Pull request by theghost5800**
JIRA:LMCROSSITXSADEPLOY-1340

### CODE REVIEW

File: com.sap.cloud.lm.sl.cf.core/src/test/java/com/sap/cloud/lm/sl/cf/core/util/SecurityUtilTest.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+            .getValue(),
+            userInfo.getToken()
+                .getValue());
+    }
+
+    public static Stream<Arguments> testGetTokenUserInfo() {
+        return Stream.of(
+        // @formatter:off
+            // (1) UserInfo with user token
+            Arguments.of("cf","CF_USER", "dUTjdafgtw3wRUMkt4XDu2IidcEHNPoh", null, new UserInfo("cf", "CF_USER", new DefaultOAuth2AccessToken("dUTjdafgtw3wRUMkt4XDu2IidcEHNPoh"))),
```
[the remark this answers is not in the document]
Comment (theghost5800) [reply com.sap.cloud.lm.sl.cf.core/src/test/java/com/sap/cloud/lm/sl/cf/core/util/SecurityUtilTest.java]: These random strings should be simillar to opaque token length

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/util/SecurityUtil.java [hunk 2]
```diff
@@ -34,7 +36,22 @@ public static OAuth2Authentication createAuthentication(String clientId, Set<Str
...
         TokenProperties tokenProperties = TokenProperties.fromToken(token);
+        OAuth2AccessToken exchangedToken = getExchangedToken(token);
+        if (exchangedToken != null) {
+            return new UserInfo(tokenProperties.getUserId(), tokenProperties.getUserName(), exchangedToken);
+        }
         return new UserInfo(tokenProperties.getUserId(), tokenProperties.getUserName(), token);
     }
 
+    private static OAuth2AccessToken getExchangedToken(OAuth2AccessToken token) {
+        String exchangedTokenValue = (String) token.getAdditionalInformation()
```
[the remark this answers is not in the document]
Comment (theghost5800) [reply com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/util/SecurityUtil.java]: Fixed. Using MapUtils to get String object safely.

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/jobs/TokensCleaner.java [hunk 3]
```diff
@@ -33,9 +35,9 @@ public TokensCleaner(@Named("tokenStore") TokenStore tokenStore) {
 
     @Override
     public void execute(Date expirationTime) {
-        Collection<OAuth2AccessToken> allTokens = tokenStore.findTokensByClientId(SecurityUtil.CLIENT_ID);
+        Collection<OAuth2AccessToken> tokensokens = tokenStore.findTokensByClientId(SecurityUtil.CLIENT_ID);
```
[the remark this answers is not in the document]
Comment (theghost5800) [reply com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/jobs/TokensCleaner.java]: fixed

File: com.sap.cloud.lm.sl.cf.web/src/main/java/com/sap/cloud/lm/sl/cf/web/configuration/UAAClientConfiguration.java [hunk 4]
```diff
@@ -51,4 +46,13 @@ private URL readTokenEndpoint(URL targetURL) {
         }
     }
 
+    protected Map<String, Object> getControllerInfo(URL targetURL) {
+        String infoURL = targetURL.toString() + "/v2/info";
```
[the remark this answers is not in the document]
Comment (theghost5800) [reply com.sap.cloud.lm.sl.cf.web/src/main/java/com/sap/cloud/lm/sl/cf/web/configuration/UAAClientConfiguration.java]: done

File: com.sap.cloud.lm.sl.cf.client/src/test/java/com/sap/cloud/lm/sl/cf/client/util/TokenFactoryTest.java [hunk 5]
```diff
@@ -99,11 +103,61 @@ public void testCreateToken() {
 
     }
 
+    @RunWith(Parameterized.class)
+    public static class ExchangedTokenFactoryTest {
```
Comment (boyan-velinov) [inline com.sap.cloud.lm.sl.cf.client/src/test/java/com/sap/cloud/lm/sl/cf/client/util/TokenFactoryTest.java]: I imagine this test to be as part of XS project or renamed to be more general for all CF use cases.

**Pull request merged**